### PR TITLE
Finalize part two

### LINKED ISSUE #76: Finalize Part-Two
- Review and flesh out all docstrings
- Ensure test coverage is adaquate
- Add problemify-style solution/problem comments for splitting staging code into stub and solution code

### PR DESCRIPTION
**Pull request by delventhalz**
~Based on #74, only the commits from `e0cf623 Remove utils modules from part-two/client` are new to this PR.~

Closes #76 

Another big PR. No real functional changes, but a lot of structural ones:
- docstrings are everywhere, and in more or less their final state
- there are problemify-style solution/problem comments throughout
- processor utils are now "services", and there is no generic "helpers" or "constants"
- added tests for the processor services
-the solution code has been rejiggered a bit to be more straight forward for students, especially in the processor

To run tests:

```bash
cd code/part-two/client
npm test
```

```bash
cd code/part-two/processor
npm test
```

To run the components:

```bash
cd code/part-two
docker-compose up
```

### CODE REVIEW

File: code/part-two/client/source/services/signing.js [hunk 1]
```diff
@@ -38,8 +54,8 @@ export const createKeys = () => {
 };
 
 /**
- * Takes a private key and returns a signing function. This function will
- * take a Buffer message, and return a hexadecimal signature.
+ * Takes a a hex private key and a string message, returning a
```
Comment (therobinkim) [inline code/part-two/client/source/services/signing.js]: typo: `a a hex private key`

File: code/part-two/processor/test/services/addressing.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  return PREFIXES.SIRE_LISTING + hash(ownerKey, 62);
+};
+
+// Takes a public key and array of moji addresses, returns an offer address
+const getOfferAddress = (ownerKey, addresses) => {
+  if (!Array.isArray(addresses)) {
+    addresses = [ addresses ];
+  }
+
+  return PREFIXES.OFFER + hash(ownerKey, 8) + hash(addresses.join(''), 54);
```
Comment (therobinkim) [inline code/part-two/processor/test/services/addressing.js]: Should we sort addresses before we join them?

File: code/part-two/processor/services/addressing.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  /* START PROBLEM
+  // Your code here
+
+  END PROBLEM */
+  // START SOLUTION
+  if (!Array.isArray(addresses)) {
+    addresses = [ addresses ];
+  }
+
+  return PREFIXES.OFFER + hash(ownerKey, 8) + hash(addresses.join(''), 54);
```
Comment (therobinkim) [inline code/part-two/processor/services/addressing.js]: Should we sort addresses before we join them?

File: code/part-two/client/source/services/addressing.js [hunk 4]
```diff
@@ -1,71 +1,138 @@
...
+
+  END PROBLEM */
+  // START SOLUTION
   if (ownerKey === null) {
-    return typePrefix;
+    return PREFIXES.OFFER;
   }
 
-  const collectionPrefix = typePrefix + hash(ownerKey, 8);
+  const collectionPrefix = PREFIXES.OFFER + hash(ownerKey, 8);
```
Comment (therobinkim) [inline code/part-two/client/source/services/addressing.js]: Should we rename `collectionPrefix` to `offerPrefix` instead?

File: code/part-two/client/source/services/addressing.js [hunk 5]
```diff
@@ -79,22 +146,9 @@ export const getOfferAddress = (ownerKey = null, moji = null) => {
       return addressOrDna;
     }
 
-    return getMojiAddress(publicKey, addressOrDna);
+    return getMojiAddress(ownerKey, addressOrDna);
   });
 
   return collectionPrefix + hash(addresses.join(''), 54);
```
Comment (therobinkim) [inline code/part-two/client/source/services/addressing.js]: Should we sort addresses before we join them?

File: code/part-two/processor/tests/services/addressing.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  return PREFIXES.SIRE_LISTING + hash(ownerKey, 62);
+};
+
+// Takes a public key and array of moji addresses, returns an offer address
+const getOfferAddress = (ownerKey, addresses) => {
+  if (!Array.isArray(addresses)) {
+    addresses = [ addresses ];
+  }
+
+  return PREFIXES.OFFER + hash(ownerKey, 8) + hash(addresses.join(''), 54);
```
Comment (therobinkim) [inline code/part-two/processor/tests/services/addressing.js]: Should we sort addresses before we join them?

File: code/part-two/processor/tests/03-CreateCollection.js [hunk 7]
```diff
@@ -31,16 +31,14 @@ describe('Create Collection', function() {
     address = getCollectionAddress(publicKey);
   });
 
-  it('should create a Collection at the correct address', function() {
+  it('should create collections at the correct address', function() {
```
Comment (therobinkim) [inline code/part-two/processor/tests/03-CreateCollection.js]: I think `address` should be pluralized too

File: code/part-two/processor/tests/03-CreateCollection.js [hunk 8]
```diff
@@ -24,19 +28,17 @@ describe('Create Collection', function() {
     context = new Context();
     txn = new Txn({ action: 'CREATE_COLLECTION' });
     publicKey = txn._publicKey;
-    address = getAddress.collection(publicKey);
+    address = getCollectionAddress(publicKey);
   });
 
-  it('should create a Collection at the correct address', function() {
+  it('should create collections at the correct address', function() {
```
Comment (therobinkim) [inline code/part-two/processor/tests/03-CreateCollection.js]: I think `address` should be plural too (`should create collections at correct addresses`)

File: code/part-two/processor/tests/09-ExtraCredit-AcceptResponse.js [hunk 9]
```diff
@@ -166,17 +163,11 @@ describe('Accept Response', function() {
...
+    const submission = handler.apply(txn, context);
 
-    return handler.apply(txn, context)
-      .catch(err => {
-        expect(err, 'Error should be an InvalidTransaction')
-          .to.be.instanceOf(InvalidTransaction);
-        return true;
-      })
+    return expect(submission).to.be.rejectedWith(InvalidTransaction)
       .then(wasRejected => {
```
Comment (therobinkim) [inline code/part-two/processor/tests/09-ExtraCredit-AcceptResponse.js]: I think `wasRejected` is no longer being used any more

File: code/part-two/client/source/services/addressing.js (hunk 5 above)
Comment (delventhalz) [reply code/part-two/client/source/services/addressing.js]: Yes. Damn. I'll add a commit.

File: code/part-two/processor/tests/07-ExtraCredit-CancelOffer.js [hunk 10]
```diff
@@ -38,7 +39,7 @@ describe('Cancel Offer', function() {
       });
   });
 
-  it('should delete a canceled offer', function() {
+  it('should delete canceled offers', function() {
```
[the remark this answers is not in the document]
Comment (delventhalz) [reply code/part-two/processor/tests/07-ExtraCredit-CancelOffer.js]: Actually either spelling is correct, and the two-L spelling is considered British.

https://www.merriam-webster.com/dictionary/cancel

File: code/part-two/processor/tests/07-ExtraCredit-CancelOffer.js [hunk 11]
```diff
@@ -47,44 +48,41 @@ describe('Cancel Offer', function() {
...
       })
       .then(wasRejected => {
         expect(wasRejected, 'Transaction should be rejected').to.be.true;
-        expect(context._state[offer], 'Offer should still exist').to.exist;
+        expect(context._state[offer]).to.exist;
       });
   });
 
-  it('should reject canceling an offer that is not set', function() {
+  it('should reject canceling offers that are not set', function() {
```
[the remark this answers is not in the document]
Comment (delventhalz) [reply code/part-two/processor/tests/07-ExtraCredit-CancelOffer.js]: Same as `canceled`.

File: code/part-two/processor/tests/09-ExtraCredit-AcceptResponse.js (hunk 9 above)
Comment (delventhalz) [reply code/part-two/processor/tests/09-ExtraCredit-AcceptResponse.js]: Commits could be out of order in the PR (GitHub does weird orders with rebases sometimes). As of this commit, `wasRejected` is still used. Later this whole messy promise construct gets refactored into a custom chai method: `rejectedWith`.

**Pull request merged**